File: Dart/Math/get_missing_integer_in_slice.dart
import "dart:core";
import "dart:math";

void main() {
  // Create a new list of numbers.
  final List<int> randomSlice = <int>[0, 9];
  // Get the missing number.
  print(getMissingIntegerInSlice(randomSlice));
}

// Find the missing int in a given slice.
List<int> getMissingIntegerInSlice(final List<int> givenSlice) {
  // Get the smallest and largest numbers in the slice.
  final int smallestInt = getSmallestNumberInSlice(givenSlice);
  final int largestInt = getBiggestNumberInSlice(givenSlice);
  // Create a slice for the missing numbers.
  final List<int> missingNumbers = <int>[];
  // Loop through the slice and add the missing numbers to the slice.
  for (int i = smallestInt; i <= largestInt; i++) {
    if (sliceCointains(givenSlice, i) == false) {
      missingNumbers.add(i);
    }
  }
  return missingNumbers;
}

// Get the smallest number in a given slice.
int getSmallestNumberInSlice(final List<int> givenSlice) {
  return givenSlice.reduce(min);
}

// Get the biggest number in a given slice.
int getBiggestNumberInSlice(final List<int> givenSlice) {
  return givenSlice.reduce(max);
}

// Check if a given slice contains a given value and return a boolean
bool sliceCointains(final List<dynamic> slice, final dynamic element) {
  return slice.contains(element);
}

File: Go/Math/get-missing-integer-in-slice.go
package main

import "fmt"

func main() {
	// Create a slice of integers
	randomInts := []int{1, 2, 4, 5}
	// Get the missing number in the slice
	fmt.Println(getMissingIntegerInSlice(randomInts))
}

// Get the missing number in a given integer array
func getMissingIntegerInSlice(slice []int) []int {
	// Get the smallest and biggest number in the slice
	smallestNumber := getSmallestNumberInSlice(slice)
	biggestNumber := getBiggestNumberInSlice(slice)
	var missingNumber []int
	for tempNumber := smallestNumber; tempNumber <= biggestNumber; tempNumber++ {
		if !sliceContains(slice, tempNumber) {
			missingNumber = append(missingNumber, tempNumber)
		}
	}
	return missingNumber
}

// Get the smallest number in a given slice.
func getSmallestNumberInSlice(slice []int) int {
	smallestNumber := slice[0]
	for _, value := range slice {
		if value < smallestNumber {
			smallestNumber = value
		}
	}
	return smallestNumber
}

// Get the biggest number in a given slice.
func getBiggestNumberInSlice(slice []int) int {
	biggestNumber := slice[0]
	for _, value := range slice {
		if value > biggestNumber {
			biggestNumber = value
		}
	}
	return biggestNumber
}

// Check if the slice contains a value and return a bool.
func sliceContains(slice []int, cointains int) bool {
	for _, value := range slice {
		if value == cointains {
			return true
		}
	}
	return false
}
